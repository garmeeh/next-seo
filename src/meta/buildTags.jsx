import React from 'react';

const defaults = {
  templateTitle: null,
  openGraph: {
    defaultImageHeight: null,
    defaultImageWidth: null,
  },
};

const buildTags = (config) => {
  const tagsToRender = [];

  if (config.titleTemplate) {
    defaults.templateTitle = config.titleTemplate;
  }

  if (config.title) {
    let updatedTitle = config.title;
    if (defaults.templateTitle) {
      updatedTitle = defaults.templateTitle.replace(/%s/g, () => updatedTitle);
    }
    tagsToRender.push(<title key="title">{updatedTitle}</title>);
  }

  if (!config.noindex) {
    tagsToRender.push(<meta key="robots" name="robots" content="index,follow" />);
    tagsToRender.push(<meta key="googlebot" name="googlebot" content="index,follow" />);
  } else {
    tagsToRender.push(<meta key="robots" name="robots" content="noindex,nofollow" />);
    tagsToRender.push(<meta key="googlebot" name="googlebot" content="noindex,nofollow" />);
  }

  if (config.description) {
    tagsToRender.push(<meta key="description" name="description" content={config.description} />);
  }

  if (config.hasOwnProperty('twitter')) {
    if (config.twitter.cardType) {
      tagsToRender.push(
        <meta key="twitter:card" name="twitter:card" content={config.twitter.cardType} />,
      );
    }

    if (config.twitter.site) {
      tagsToRender.push(
        <meta key="twitter:site" name="twitter:site" content={config.twitter.site} />,
      );
    }

    if (config.twitter.handle) {
      tagsToRender.push(
        <meta key="twitter:creator" name="twitter:creator" content={config.twitter.handle} />,
      );
    }
  }

  if (config.hasOwnProperty('openGraph')) {
    if (config.openGraph.url) {
      tagsToRender.push(<meta key="og:url" property="og:url" content={config.openGraph.url} />);
    }

    if (config.openGraph.type) {
      tagsToRender.push(<meta key="og:type" property="og:type" content={config.openGraph.type} />);
    }

    if (config.openGraph.title) {
      tagsToRender.push(
        <meta key="og:title" property="og:title" content={config.openGraph.title} />,
      );
    }

    if (config.openGraph.description) {
      tagsToRender.push(
        <meta
          key="og:description"
          property="og:description"
          content={config.openGraph.description}
        />,
      );
    }

    if (config.openGraph.images && config.openGraph.images.length) {
      config.openGraph.images.forEach((image, index) => {
        tagsToRender.push(
          <meta key={`og:image:0${index}`} property="og:image" content={image.url} />,
        );

        if (image.alt) {
          tagsToRender.push(
            <meta key={`og:image:alt0${index}`} property="og:image:alt" content={image.alt} />,
          );
        }

        if (image.width) {
          tagsToRender.push(
            <meta
              key={`og:image:width0${index}`}
              property="og:image:width"
              content={image.width}
            />,
          );
        } else if (defaults.openGraph.defaultImageWidth || config.openGraph.defaultImageWidth) {
          if (config.openGraph.defaultImageWidth) {
            defaults.openGraph.defaultImageWidth = config.openGraph.defaultImageWidth;
          }
          tagsToRender.push(
            <meta
              key={`og:image:width0${index}`}
              property="og:image:width"
              content={defaults.openGraph.defaultImageWidth}
            />,
          );
        }

        if (image.height) {
          tagsToRender.push(
            <meta
              key={`og:image:height${index}`}
              property="og:image:height"
              content={image.height}
            />,
          );
        } else if (defaults.openGraph.defaultImageHeight || config.openGraph.defaultImageHeight) {
          if (config.openGraph.defaultImageHeight) {
            defaults.openGraph.defaultImageHeight = config.openGraph.defaultImageHeight;
          }
          tagsToRender.push(
            <meta
              key={`og:image:height${index}`}
              property="og:image:height"
              content={defaults.openGraph.defaultImageHeight}
            />,
          );
        }
      });
    }

    if (config.openGraph.locale) {
      tagsToRender.push(
        <meta key="og:locale" property="og:locale" content={config.openGraph.locale} />,
      );
    }

    if (config.openGraph.locale) {
      tagsToRender.push(
        <meta key="og:locale" property="og:locale" content={config.openGraph.locale} />,
      );
    }

    if (config.openGraph.locale) {
      tagsToRender.push(
        <meta key="og:locale" property="og:locale" content={config.openGraph.locale} />,
      );
    }

    if (config.openGraph.site_name) {
      tagsToRender.push(
        <meta key="og:site_name" property="og:site_name" content={config.openGraph.site_name} />,
      );
    }
  }

  if (config.canonical) {
    tagsToRender.push(<link rel="canonical" href={config.canonical} key="canonical" />);
  }

  return tagsToRender;
};

export default buildTags;
