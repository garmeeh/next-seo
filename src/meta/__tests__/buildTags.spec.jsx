import { getByText, render } from 'react-testing-library';
import React from 'react';

import buildTags from '../buildTags';

const SEO = {
  title: 'This is a test title.',
  description: 'This is a test description.',
  canonical: 'https://www.canonical.ie',
  openGraph: {
    type: 'website',
    locale: 'en_IE',
    url: 'https://www.url.ie',
    title: 'Open graph title',
    description: 'This is testing og:description.',
    defaultImageWidth: 1200,
    defaultImageHeight: 1200,
    images: [
      {
        url: 'https://www.test.ie/image-01.jpg',
        width: 800,
        height: 600,
        alt: 'Alt text right here',
      },
      { url: 'https://www.test.ie/image-02.jpg' },
      { url: 'https://www.test.ie/image-03.jpg' },
      { url: 'https://www.test.ie/image-04.jpg' },
    ],
    site_name: 'SiteName',
  },
  twitter: {
    handle: '@handle',
    site: '@site',
    cardType: 'summary_large_image',
  },
};

it('renders correctly', () => {
  const tags = buildTags(SEO);
  const { container } = render(tags);
  expect(container).toMatchSnapshot();
});

/* eslint no-magic-numbers: 0 */
it('returns full array for default seo object', () => {
  const tags = buildTags(SEO);
  const { container } = render(tags);

  const title = getByText(
    container,
    (content, element) =>
      element.tagName.toLowerCase() === 'title' &&
      content.startsWith(`${SEO.title}`),
  );
  const index = container.querySelectorAll('meta[content="index,follow"]');
  const description = container.querySelectorAll(
    `meta[content="${SEO.description}"]`,
  );
  const descriptionTag = container.querySelectorAll('meta[name="description"]');
  const twitterCard = container.querySelectorAll(
    'meta[content="summary_large_image"]',
  );
  const twitterCardTag = container.querySelectorAll(
    'meta[name="twitter:card"]',
  );
  const twitterHandle = container.querySelectorAll(
    `meta[content="${SEO.twitter.handle}"]`,
  );
  const twitterHandleTag = container.querySelectorAll(
    'meta[name="twitter:creator"]',
  );
  const twitterSite = container.querySelectorAll(
    `meta[content="${SEO.twitter.site}"]`,
  );
  const twitterSiteTag = container.querySelectorAll(
    'meta[name="twitter:site"]',
  );
  const ogUrl = container.querySelectorAll(
    `meta[content="${SEO.openGraph.url}"]`,
  );
  const ogUrlTag = container.querySelectorAll('meta[property="og:url"]');
  const ogType = container.querySelectorAll(
    `meta[content="${SEO.openGraph.type}"]`,
  );
  const ogTypeTag = container.querySelectorAll('meta[property="og:type"]');
  const ogTitle = container.querySelectorAll(
    `meta[content="${SEO.openGraph.title}"]`,
  );
  const ogTitleTag = container.querySelectorAll('meta[property="og:title"]');
  const ogDescription = container.querySelectorAll(
    `meta[content="${SEO.openGraph.description}"]`,
  );
  const ogDescriptionTag = container.querySelectorAll(
    'meta[property="og:description"]',
  );
  const ogImage00 = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[0].url}"]`,
  );
  const ogImageTag00 = tags.filter(item => item.key === 'og:image:01');
  const ogImage01 = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[1].url}"]`,
  );
  const ogImageTag01 = tags.filter(item => item.key === 'og:image:01');
  const ogImage02 = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[2].url}"]`,
  );
  const ogImageTag02 = tags.filter(item => item.key === 'og:image:02');
  const ogImage03 = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[3].url}"]`,
  );
  const ogImageTag03 = tags.filter(item => item.key === 'og:image:03');
  const ogDefaultImageWidthHeight = container.querySelectorAll(
    `meta[content="${SEO.openGraph.defaultImageHeight}"]`,
  );
  const ogSetImageHeight = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[0].height}"]`,
  );
  const ogSetImageWidth = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[0].width}"]`,
  );
  const ogSetImageAlt = container.querySelectorAll(
    `meta[content="${SEO.openGraph.images[0].alt}"]`,
  );
  const ogLocale = container.querySelectorAll(
    `meta[content="${SEO.openGraph.locale}"]`,
  );
  const ogLocaleTag = tags.filter(item => item.key === 'og:locale');
  const ogSiteName = container.querySelectorAll(
    `meta[content="${SEO.openGraph.site_name}"]`,
  );
  const ogSiteNameTag = tags.filter(item => item.key === 'og:site_name');
  const canonicalTag = tags.filter(item => item.key === 'canonical');

  expect(title).toBeDefined();
  expect(Array.from(index).length).toBe(2);
  expect(Array.from(description).length).toBe(1);
  expect(Array.from(descriptionTag).length).toBe(1);
  expect(Array.from(twitterCard).length).toBe(1);
  expect(Array.from(twitterCardTag).length).toBe(1);
  expect(Array.from(twitterHandle).length).toBe(1);
  expect(Array.from(twitterHandleTag).length).toBe(1);
  expect(Array.from(twitterSite).length).toBe(1);
  expect(Array.from(twitterSiteTag).length).toBe(1);
  expect(Array.from(ogUrl).length).toBe(1);
  expect(Array.from(ogUrlTag).length).toBe(1);
  expect(Array.from(ogType).length).toBe(1);
  expect(Array.from(ogTypeTag).length).toBe(1);
  expect(Array.from(ogTitle).length).toBe(1);
  expect(Array.from(ogTitleTag).length).toBe(1);
  expect(Array.from(ogDescription).length).toBe(1);
  expect(Array.from(ogDescriptionTag).length).toBe(1);
  expect(Array.from(ogImage00).length).toBe(1);
  expect(Array.from(ogImageTag00).length).toBe(1);
  expect(Array.from(ogImage01).length).toBe(1);
  expect(Array.from(ogImageTag01).length).toBe(1);
  expect(Array.from(ogImage02).length).toBe(1);
  expect(Array.from(ogImageTag02).length).toBe(1);
  expect(Array.from(ogImage03).length).toBe(1);
  expect(Array.from(ogImageTag03).length).toBe(1);
  expect(Array.from(ogDefaultImageWidthHeight).length).toBe(6);
  expect(Array.from(ogSetImageHeight).length).toBe(1);
  expect(Array.from(ogSetImageWidth).length).toBe(1);
  expect(Array.from(ogSetImageAlt).length).toBe(1);
  expect(Array.from(ogLocale).length).toBe(1);
  expect(Array.from(ogLocaleTag).length).toBe(1);
  expect(Array.from(ogSiteName).length).toBe(1);
  expect(Array.from(ogSiteNameTag).length).toBe(1);
  expect(canonicalTag[0].props.href).toBe(`${SEO.canonical}`);
  expect(Array.from(canonicalTag).length).toBe(1);
});

it('correctly sets noindex, nofollow', () => {
  const overrideProps = {
    ...SEO,
    noindex: true,
  };
  const tags = buildTags(overrideProps);
  const { container } = render(tags);
  const index = container.querySelectorAll('meta[content="index,follow"]');
  const noindex = container.querySelectorAll(
    'meta[content="noindex,nofollow"]',
  );

  expect(Array.from(index).length).toBe(0);
  expect(Array.from(noindex).length).toBe(2);
});

it('displays title with titleTemplate integrated', () => {
  const template = 'Next SEO';
  const overrideProps = {
    ...SEO,
    titleTemplate: `${template} | %s`,
  };
  const tags = buildTags(overrideProps);
  const { container } = render(tags);
  const title = getByText(
    container,
    (content, element) =>
      element.tagName.toLowerCase() === 'title' && content.startsWith(template),
  );
  expect(title.innerHTML).toMatch(`${template} | ${SEO.title}`);
});
