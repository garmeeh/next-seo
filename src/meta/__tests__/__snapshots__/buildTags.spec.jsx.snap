// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article SEO renders correctly 1`] = `
<div>
  <title>
    Next SEO | Article Page Title
  </title>
  <meta
    content="index,follow"
    name="robots"
  />
  <meta
    content="index,follow"
    name="googlebot"
  />
  <meta
    content="Description of article page"
    name="description"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@site"
    name="twitter:site"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="https://www.example.com/articles/article-title"
    property="og:url"
  />
  <meta
    content="article"
    property="og:type"
  />
  <meta
    content="2017-06-21T23:04:13Z"
    property="article:published_time"
  />
  <meta
    content="2018-01-21T18:04:43Z"
    property="article:modified_time"
  />
  <meta
    content="2022-12-21T22:04:11Z"
    property="article:expiration_time"
  />
  <meta
    content="https://www.example.com/authors/@firstnameA-lastnameA"
    property="article:author"
  />
  <meta
    content="https://www.example.com/authors/@firstnameB-lastnameB"
    property="article:author"
  />
  <meta
    content="Section II"
    property="article:section"
  />
  <meta
    content="Tag A"
    property="article:tag"
  />
  <meta
    content="Tag B"
    property="article:tag"
  />
  <meta
    content="Open Graph Article Title"
    property="og:title"
  />
  <meta
    content="Description of open graph article"
    property="og:description"
  />
  <meta
    content="https://www.test.ie/og-image-article-title-01.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt Article Title A"
    property="og:image:alt"
  />
  <meta
    content="850"
    property="og:image:width"
  />
  <meta
    content="650"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/og-image-article-title-02.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt Article Title B"
    property="og:image:alt"
  />
  <meta
    content="950"
    property="og:image:width"
  />
  <meta
    content="850"
    property="og:image:height"
  />
  <meta
    content="SiteName"
    property="og:site_name"
  />
</div>
`;

exports[`Book SEO renders correctly 1`] = `
<div>
  <title>
    Next SEO | Book Page Title
  </title>
  <meta
    content="index,follow"
    name="robots"
  />
  <meta
    content="index,follow"
    name="googlebot"
  />
  <meta
    content="Description of book page"
    name="description"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@site"
    name="twitter:site"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="https://www.example.com/books/book-title"
    property="og:url"
  />
  <meta
    content="book"
    property="og:type"
  />
  <meta
    content="https://www.example.com/authors/@firstnameA-lastnameA"
    property="book:author"
  />
  <meta
    content="https://www.example.com/authors/@firstnameB-lastnameB"
    property="book:author"
  />
  <meta
    content="978-3-16-148410-0"
    property="book:isbn"
  />
  <meta
    content="2018-09-17T11:08:13Z"
    property="book:release_date"
  />
  <meta
    content="Tag A"
    property="book:tag"
  />
  <meta
    content="Tag B"
    property="book:tag"
  />
  <meta
    content="Open Graph Book Title"
    property="og:title"
  />
  <meta
    content="Description of open graph book"
    property="og:description"
  />
  <meta
    content="https://www.test.ie/og-image-book-title-01.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt Book Title A"
    property="og:image:alt"
  />
  <meta
    content="850"
    property="og:image:width"
  />
  <meta
    content="650"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/og-image-book-title-02.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt Book Title B"
    property="og:image:alt"
  />
  <meta
    content="950"
    property="og:image:width"
  />
  <meta
    content="850"
    property="og:image:height"
  />
  <meta
    content="SiteName"
    property="og:site_name"
  />
</div>
`;

exports[`Profile SEO renders correctly 1`] = `
<div>
  <title>
    Next SEO | Profile Page Title
  </title>
  <meta
    content="index,follow"
    name="robots"
  />
  <meta
    content="index,follow"
    name="googlebot"
  />
  <meta
    content="Description of profile page"
    name="description"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@site"
    name="twitter:site"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="https://www.example.com/@firstlast123"
    property="og:url"
  />
  <meta
    content="profile"
    property="og:type"
  />
  <meta
    content="First"
    property="profile:first_name"
  />
  <meta
    content="Last"
    property="profile:last_name"
  />
  <meta
    content="firstlast123"
    property="profile:username"
  />
  <meta
    content="male"
    property="profile:gender"
  />
  <meta
    content="Open Graph Profile Title"
    property="og:title"
  />
  <meta
    content="Description of open graph profile"
    property="og:description"
  />
  <meta
    content="https://www.test.ie/og-image-firstlast123-01.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt firstlast123 A"
    property="og:image:alt"
  />
  <meta
    content="850"
    property="og:image:width"
  />
  <meta
    content="650"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/og-image-firstlast123-02.jpg"
    property="og:image"
  />
  <meta
    content="Og Image Alt firstlast123 B"
    property="og:image:alt"
  />
  <meta
    content="950"
    property="og:image:width"
  />
  <meta
    content="850"
    property="og:image:height"
  />
  <meta
    content="SiteName"
    property="og:site_name"
  />
</div>
`;

exports[`renders correctly 1`] = `
<div>
  <title>
    This is a test title.
  </title>
  <meta
    content="index,follow"
    name="robots"
  />
  <meta
    content="index,follow"
    name="googlebot"
  />
  <meta
    content="This is a test description."
    name="description"
  />
  <meta
    content="summary_large_image"
    name="twitter:card"
  />
  <meta
    content="@site"
    name="twitter:site"
  />
  <meta
    content="@handle"
    name="twitter:creator"
  />
  <meta
    content="1234567890"
    name="fb:app_id"
  />
  <meta
    content="https://www.url.ie"
    property="og:url"
  />
  <meta
    content="website"
    property="og:type"
  />
  <meta
    content="Open graph title"
    property="og:title"
  />
  <meta
    content="This is testing og:description."
    property="og:description"
  />
  <meta
    content="https://www.test.ie/image-01.jpg"
    property="og:image"
  />
  <meta
    content="Alt text right here"
    property="og:image:alt"
  />
  <meta
    content="800"
    property="og:image:width"
  />
  <meta
    content="600"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/image-02.jpg"
    property="og:image"
  />
  <meta
    content="1200"
    property="og:image:width"
  />
  <meta
    content="1200"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/image-03.jpg"
    property="og:image"
  />
  <meta
    content="1200"
    property="og:image:width"
  />
  <meta
    content="1200"
    property="og:image:height"
  />
  <meta
    content="https://www.test.ie/image-04.jpg"
    property="og:image"
  />
  <meta
    content="1200"
    property="og:image:width"
  />
  <meta
    content="1200"
    property="og:image:height"
  />
  <meta
    content="en_IE"
    property="og:locale"
  />
  <meta
    content="SiteName"
    property="og:site_name"
  />
  <link
    href="https://www.canonical.ie"
    rel="canonical"
  />
</div>
`;
